pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,289 @@
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/ceohelpers"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+// machineDeletionHookName holds a name of the Machine Deletion Hook
+const machineDeletionHookName = "EtcdQuorumOperator"
+
+// machineDeletionHookOwner holds an owner of the Machine Deletion Hook
+const machineDeletionHookOwner = "clusteroperator/etcd"
+
+type machineDeletionHooksController struct {
+	etcdClient                            etcdcli.EtcdClient
+	machineClient                         machinev1beta1client.MachineInterface
+	masterMachineLister                   machinelistersv1beta1.MachineLister
+	configMapListerForKubeSystemNamespace corev1listers.ConfigMapNamespaceLister
+	// machineAPIChecker determines if the precondition for this controller is met,
+	// this controller can be run only on a cluster that exposes a functional Machine API
+	machineAPIChecker     ceohelpers.MachineAPIChecker
+	masterMachineSelector labels.Selector
+}
+
+// NewMachineDeletionHooksController reconciles machine hooks for master machines
+//
+// note:
+//  The Machine Deletion Hook this controller reconciles is a mechanism within the Machine API that allow this operator to hold up removal of a machine
+//  until a replacement member has been promoted to a voting member.
+func NewMachineDeletionHooksController(
+	operatorClient operatorv1helpers.OperatorClient,
+	machineClient machinev1beta1client.MachineInterface,
+	etcdClient etcdcli.EtcdClient,
+	machineAPIChecker ceohelpers.MachineAPIChecker,
+	masterMachineSelector labels.Selector,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	masterMachineInformer cache.SharedIndexInformer,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &machineDeletionHooksController{
+		machineClient:                         machineClient,
+		etcdClient:                            etcdClient,
+		machineAPIChecker:                     machineAPIChecker,
+		masterMachineSelector:                 masterMachineSelector,
+		masterMachineLister:                   machinelistersv1beta1.NewMachineLister(masterMachineInformer.GetIndexer()),
+		configMapListerForKubeSystemNamespace: kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Lister().ConfigMaps("kube-system"),
+	}
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(
+			kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Informer(),
```

Comment:
Are we only re-syncing on `kube-system` configmaps?
I would have thought this controller would need to react to changes on Machines for hook changes, and potentially nodes or etcd pods to detect member addition/removal.

Reply:
good point, added `masterMachineInformer` and changed the resync time to 1 minute